x-pack/solutions/security/plugins/security_solution/server/lib/entity_analytics/risk_engine/migrations/rename_risk_score_component_templates.ts

```diff
@@ -0,0 +1,83 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { asyncForEach } from '@kbn/std';
+import type { EntityAnalyticsMigrationsParams } from '../../migrations';
+import { RiskScoreDataClient } from '../../risk_score/risk_score_data_client';
+import type { RiskEngineConfiguration } from '../../types';
+import { riskEngineConfigurationTypeName } from '../saved_object';
+import { buildScopedInternalSavedObjectsClientUnsafe } from '../../risk_score/tasks/helpers';
+import { mappingComponentName } from '../../risk_score/configurations';
+
+export const MAX_PER_PAGE = 10_000;
+
+/**
+ * This migration renames the Risk Score component templates to include the namespace in the name. Before 8.18 all spaces used the `.risk-score-mappings` component template, we now use `.risk-score-mappings-<spacename>`.
+ *
+ * The migration creates the new component template and updates the index template for each space, then finally deletes the old component template.
+ */
+export const renameRiskScoreComponentTemplate = async ({
+  auditLogger,
+  logger,
+  getStartServices,
+  kibanaVersion,
+}: EntityAnalyticsMigrationsParams) => {
+  const [coreStart] = await getStartServices();
+  const soClientKibanaUser = coreStart.savedObjects.createInternalRepository();
+  const esClient = coreStart.elasticsearch.client.asInternalUser;
+
+  // Check if the legacy component templates (without the namespace in the name) exists
+  const oldComponentTemplateExists = await esClient.cluster.existsComponentTemplate({
+    name: mappingComponentName,
+  });
+
+  if (!oldComponentTemplateExists) {
+    return;
+  }
+
+  // Get all installed Risk Engine Configurations
+  const savedObjectsResponse = await soClientKibanaUser.find<RiskEngineConfiguration>({
+    type: riskEngineConfigurationTypeName,
+    perPage: MAX_PER_PAGE,
+    namespaces: ['*'],
+  });
+
+  await asyncForEach(savedObjectsResponse.saved_objects, async (savedObject) => {
+    const namespace = savedObject.namespaces?.[0]; // We need to create one component template per space
+
+    if (!namespace) {
+      logger.error('Unexpected saved object. Risk Score saved objects must have a namespace');
+      return;
+    }
+
+    logger.info(`Starting Risk Score component template migration on namespace ${namespace}`);
+
+    const soClient = buildScopedInternalSavedObjectsClientUnsafe({ coreStart, namespace });
+
+    const riskScoreDataClient = new RiskScoreDataClient({
+      logger,
+      kibanaVersion,
+      esClient,
+      namespace,
+      soClient,
+      auditLogger,
+    });
+
+    await riskScoreDataClient.createOrUpdateRiskScoreComponentTemplate();
+    await riskScoreDataClient.createOrUpdateRiskScoreIndexTemplate();
+
+    logger.debug(`Risk score component template migration ran on namespace ${namespace}`);
+  });
+
+  // Delete the legacy component template without the namespace in the name
+  await esClient.cluster.deleteComponentTemplate(
```

Comment:
I think we should just logger.error in the case that this delete fails as it is not catastrophic enough to stop kibana startup.

Reply:
Hmm, do you mean if it fails because ES is unavailable? It already ignores the 404 error.

Reply:
Maybe I'm being over cautious 👍  I can't think of another error other than ES being unavailable, which Kibana already handles on startup

Reply:
Migration failures won't stop kibana. This code "ignores" them:

https://github.com/elastic/kibana/blob/649451c6f299b1fa646313bcef8ec07fa71f8a73/x-pack/solutions/security/plugins/security_solution/server/plugin.ts#L239-L247

Reply:
Ah I didn't know! Thanks Pablo 👍